### ISSUE TITLE
"Expected N+1 tests, received N" caused by fun with @TestTemplate

### ISSUE THREAD

**Issue by mgroth0** [mannodermaus/android-junit5#316]
I was getting "Expected N+1 tests, received N" errors, and at first it was very difficult to find the cause. After a thorough investigation I found that:

- The "Expected N+1 tests, received N" error was only happening for android instrumented Junit5 tests. 
- Did not occur for regular JVM Junit5 tests or android Junit 5 unit tests 
- Only occured when a certain top-level function was on the classpath with a signature like:

```
@TestTemplate fun someFunction() {}
```

If I comment out `@TestTemplate` above, the problem goes away.

For context, note that `someFunction` above was just some scrap code I through somewhere when I was messing around and forget to get rid of. It is located in a common test dependency that is shared by JVM desktop, android unit tests, and android instrumented tests. It never caused any issue until I set up `android-junit5` for my instrumented tests.

**Comment by mannodermaus** [comment]
Hey, thanks for reporting. First of all, I'd like to ask what version of JUnit 5, the android-junit5 plugin and the instrumentation libraries you are using in this project.

I'm trying to repro your case but test template functions are discovered and executed just fine. For completeness sake, did your scrap function _only_ have the `@TestTemplate` on it, or did you omit something for brevity? The thing is, a test template is supposed to be used with some other annotation that provides context to what the template is generating.

```kotlin
// Don't
@TestTemplate
fun myTemplate1() {}

// Do
@TestTemplate
@ExtendWith(SomeTemplateContextProvider::class)
fun myTemplate2(param: SomeTemplateContext) {}
```

If you are missing that context provider, maybe this is the reason why you were seeing the miscount of the test engine? I'm not getting a miscount on the latest dev version of the libraries, even when I purposefully misconfigure the template declaration, so it's possible that I unknowingly fixed something in the meantime.

**Comment by mannodermaus** [comment]
Thank you for the thorough reproducer, this was super helpful in getting to the bottom of this. Your usage of Kotlin Multiplatform had me worried for a bit, but it turns out that it doesn't have any say in the issue appearing. After a few hours of tracing and digging, I was finally able to pinpoint the culprit and [fix it](https://github.com/mannodermaus/android-junit5/pull/325/files#diff-4fe4e0d302089f13767fac22e370f2465fed13ae4fe5c90be25ce4c707b10140R42-R43). Some new unit tests make sure that top-level functions won't be prematurely considered by the JUnit 5 builder going forward, preventing a mismatch between what we report to the Android instrumentation at the start, and what JUnit 5 ends up actually executing.

This will be available in the upcoming `1.5.0` release of the instrumentation libraries (the `runner` artifact, to be precise). If you need the fix right now, you can use the `1.5.0-SNAPSHOT` instead. Cheers!
